### DISCUSSION TITLE
AsyncSession selectinload fails in some situation

### DISCUSSION THREAD

**Discussion by mockingjet** [sqlalchemy/sqlalchemy#7057] [Usage Questions] [answered]
It's hard for me to reproduce this error because it requires lots of tables / relationships (production code). I try to explain every line detailedly.

```py

# services.py

class ProviderCategoryService(
    BaseService[ProviderCategory, ProviderCategoryCreate, ProviderCategoryUpdate]
):
    async def get_matched_providers(self, area: str, domain: str) -> list[Provider]:
        query = (
            sa.select(ProviderCategory)
            .where(
                ProviderCategory.patent_area == area,
                ProviderCategory.patent_domain == domain,
            )
            .options(sa.orm.selectinload(ProviderCategory.provider))
        )

        result = await self.db.execute(query)
        return list(map(lambda c: c.provider, result.scalars().all()))


class ProviderService(BaseService[Provider, ProviderCreate, ProviderUpdate]):
    async def get_with_user(self, provider_id: str) -> Provider:
        query = (
            sa.select(Provider)
            .where(Provider.id == provider_id)
            .options(sa.orm.selectinload(Provider.user))
        )

        result = await self.db.execute(query)
        return result.scalars().first()
```

```py

# api.py

@router.post(
    "/self/projects/{project_id}/publish",
    response_model=APIResponse[None],
)
async def publish_project(
    bt: BackgroundTasks,
    project_id: int = Path(...),
    current_user: UserRead = Depends(get_current_user),
    user_service: UserService = Depends(get_user_service),
    provider_service: ProviderService = Depends(get_provider_service),
    project_service: ProjectService = Depends(get_project_service),
    provider_category_service: ProviderCategoryService = Depends(
        get_provider_category_service
    ),
    match_service: MatchService = Depends(get_match_service),
):
    # simply get a project instance
    project = await project_service.get(project_id)

    if not project:
        raise HTTPException(404)

    if str(project.user_id) != str(current_user.id):
        raise HTTPException(403)
    
    # just modify project.status
    await project_service.publish(project)
   
    # provider_categories is a N - N - N table (provider, area, domain)
    # so based on this table, I can filter out providers by patent_area and patent_domain
    # see services.py above
    matched_providers = await provider_category_service.get_matched_providers(
        project.patent_area,
        project.patent_domain,
    )
    
    # match is a N-N relationship table for projects and providers
    matches = await match_service.create_list(project.id, list(map(lambda p: p.id, matched_providers)))

    # What confuses me is here 
    for match in matches:
        # see services.py above
        # here I try to get provider with user attribute (one to one relationship)

        # class Provider(Base, TimestampMixin, ManagerReviewedMixin):
        #     id = sa.Column(sa.ForeignKey("users.id"), primary_key=True)
        #     ...
        #     user = sa.orm.relationship(
        #         "User",
        #         uselist=False,
        #         lazy="noload",
        #     )
        
        # but provider_service.get_with_user doesn't selectinload user correctly
        match.provider = await provider_service.get_with_user(match.provider_id)
        print(match.provider.user) # None

        # So I have to manually query user to add into provider object
        match.provider.user = await user_service.get(match.provider_id)
        bt.add_task(send_match_email, match)

    return {
        "code": 0,
        "data": None,
    }
```

Forgive me if you think I don't' address my question clearly. I can give further information if you need.

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's expected, since provider is already in the session since it was loaded by `provider_category_service.get_matched_providers`.
You can refer to this discussion, that is about another problem, but the solution is the same https://github.com/sqlalchemy/sqlalchemy/discussions/6989

**Reply by mockingjet** [reply]
I have the feeling that **provider** is being "kidnapped" with `provider_category_service.get_matched_providers`. That's why I can't selectinload it later. **populate_existing** once came into my mind, but it's an unfamiliar concept for me, so I skipped this option which is exactly the solution.

> The populate_existing execution option ensures that for all rows loaded, the corresponding instances in the Session will be fully refreshed, erasing any existing data within the objects (including pending changes) and replacing with the data loaded from the result.

Does this mean that, when objects are already in the session, and being made some changes, all objects will be loaded without previous changes with **populate_existing**?

I don't know the differences (side effects) that using `populate_existing` will cause. Can you add some concrete examples in the docs?

**Reply by CaselIT** [reply]
The session should autoflush if there are changes before doing any other db access. This is controlled by https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=autoflush#sqlalchemy.orm.Session.params.autoflush

See also https://docs.sqlalchemy.org/en/14/orm/session_basics.html#flushing

You can of course also make it explicit by adding `await session.flush()`

**Reply by mockingjet** [reply]
I know the usage of `session.flush()`. And I actually use it in `await match_service.create_list` so no problems here. My point is if there is no side effect with populate_existing, why not make it as default option?

**Reply by CaselIT** [reply]
I think @zzzeek is the best person to answer that

**Reply by zzzeek** [reply]
populate existing overwrites all python attributes with the value that was loaded from the database, regardless of whether or not those attributes have pending changes on them.  therefore this option should only be used either with autoflush, or calling flush() beforehand.  was that the question?

**Reply by mockingjet** [reply]
```py
# database.py

async_session: Callable[[], AsyncSession] = sessionmaker(
    async_engine,
    expire_on_commit=False,
    class_=AsyncSession,
)

print(async_session.__dict__)
# {
#     'kw': {
#         'bind': <sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x7f78baef6270>,
#         'autoflush': True,
#         'autocommit': False,
#         'expire_on_commit': False
#     },
#    'class_': <class 'sqlalchemy.orm.session.AsyncSession'>
# }
```

My async_session has `autoflush=True`, then is there a way to make `populate_existing=True` as default select option?

I feel that **loading relational objects needs to explicitly declare populate_existing=True** kind of counter-intuitive.

**Reply by zzzeek** [reply]
You can set up populate existing for all queries using the do_orm_execute() event hook.  I'm on my phone right now so can't write out an example.  

That said there is no "need" for relationship loads to use populate existing unless you are trying to reload the same relationship in different ways, like contains_eager().   Your example at the top doesn't really show me whats really happening but pop existing is not normally required .

**Reply by zzzeek** [reply]
I see you are using noload, that's probably where things are going wrong.  There's no need to use this option either - it basically means the relationship will have the wrong data inside of the collection (ie nothing when there should be something).  I would advise using raiseload instead and make sure this attribute is not accessed for use cases where access is not expected.

**Reply by mockingjet** [reply]
Using noload is a way to handling Fastapi Pydantic Response Model, prevent it from accessing unloaded attribute and throwing greenlet.spawn error

I don't think it's about using noload. I tried using selectin and the issue is still there.

**Reply by zzzeek** [reply]
i would need a real MCVE to see what you're doing.

**Reply by zzzeek** [reply]
i also think pydantic should have a means of controlling the scope which it accesses on an object because putting "noload" means you are having the object return incorrect data.

**Reply by zzzeek** [reply]
this seems pretty clear:
```
 # class Provider(Base, TimestampMixin, ManagerReviewedMixin):
        #     id = sa.Column(sa.ForeignKey("users.id"), primary_key=True)
        #     ...
        #     user = sa.orm.relationship(
        #         "User",
        #         uselist=False,
        #         lazy="noload",
        #     )
        
        # but provider_service.get_with_user doesn't selectinload user correctly
        match.provider = await provider_service.get_with_user(match.provider_id)
        print(match.provider.user) # None
```

user is None due to "noload".  if you take out that "lazy='noload'" and change it to "selectinload" ("joinedload" a better choice for many-to-one though), user will not be None provided there's a value in the database. is that what you're seeing?

**Reply by mockingjet** [reply]
See services.py above, I explicitly select in load with user in `provider_service.get_with_user`. In `class Provider`, I set it default to noload 

As for your statement:

>putting "noload" means you are having the object return incorrect data.

Well... I didn't see related descriptions in [docs](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.noload), perhaps I missed it elsewhere?

But here it says nolaod simply does **accessed without producing any loading effect**

**Reply by zzzeek** [reply]
the User already got loaded, "noload" did its job and set that attribute to blank, as the [docs](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html) say "that will never load or have any loading effect. "    .  the original purpose of "noload" was a recipe where there's a list there that one can write towards to affect the flush, but that's it; reading the attribute was supposed to be meaningless.

the data here is "incorrect" because it's returning None, which is not the correct value of "provider.user".  it's incorrect data.

"noload" was not meant to be a means of controlling serialization by a pydantic object.   IMO if serialization wants only attributes A, B, C and not D, it should explicitly ask for A, B, C, and not D.  getting D that has "None" stuffed into it is kind of hacky.

**Reply by zzzeek** [reply]
im not claiming there can't be some feature for this use case, but we dont really have this right now.   you can approximate it with noload but you have to reset that attribute if you want it to actually load using session.expire() or populate_existing.

**Reply by mockingjet** [reply]
I see your point about "noload". Thanks for pointing out my incorrect concept.

To reproduce this issue may take me a lot of time. I won't do it for now, but will be back to this issue with a code example some day.

Thanks you both.

**Reply by zzzeek** [reply]
no need for an example i see what the issue is

**Reply by CaselIT** [reply]
Personally I think this is a pendatic issue or at least a missing option. I don't have a lot of experience with it but IIRC when serialising you can list some attributes to exclude.

I'm not sure if additional documentation is usefulz since there is already quite a bit. Maybe we could just mention expired/populate_existing in the no load docs.

Alternatively would a loader that behaves like noload on loading, but sets the attributes as expired  when reloading (so that expire is not required) be difficult to implement?

**Reply by zzzeek** [reply]
I think noload needs more caveats because we do get people using it for things and running into problems.

the thing that people "want" is what you mention, a "noload" that doesn't actually populate the attribute.  we do have the concept of attributes that return a default value when accessed, but don't populate; this occurs for transient and pending objects.  getting this to work for persistent objects is feasible but it's going to open up a lot of test cases.  these loader strategy things have lots and lots of permutations so they are hard to test fully.

**Reply by zzzeek** [reply]
an easier way might be some version of "noload" that just does what it does but also adds some flag to the state indicating "use populate existing for this", something like that.

there's also the notion that we would support a session-level option to default populate_existing to True.   the reason populate_existing was never a default behavior is because we used to not have autoflush.  if you have autoflush turned on, populate_existing in all cases is *probably* safe-ish.   i've seen this done in some openstack cases for example.

**Reply by CaselIT** [reply]
> an easier way might be some version of "noload" that just does what it does but also adds some flag to the state indicating "use populate existing for this", something like that.

That was what I was thinking of, a new attribute flag that's not 'loaded'|'expired' but 'loaded_but_reload_on_query'

**Reply by zzzeek** [reply]
we could likely add execution_options to the Session constructor.   I just checked and at the moment populate_existing has to be set up at the Session level , it wont work if applied to the engine or connection.
